streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -16,6 +16,7 @@
  */
 package org.apache.kafka.streams.processor.internals;
 
+import java.util.TreeMap;
```

Comment:
Did not go through this class in this pass, please lmk if there are some logic that needs review beyond the revoked / owned modification

Reply:
Actually not too much, but to sum it up:

- Don't encode active tasks in COOPERATIVE, just work them out from the encoded `ownedPartitions`
- If any subscription has non-empty `ownedPartitions` follow COOPERATIVE protocol (this can happen during upgrade)
- Keep track of all partitions claimed as owned and their previous owner. When assigning tasks to consumers, check if the consumer owned this partition before. If it did not but someone else owned it, don't remove that partition/active task from the assignment

There is definitely room for improvement/optimization in the assigning of tasks to consumers within a client (modified it only slightly so that if nothing changes the task->consumer mapping should not either) but that can wait